src/include/server/mir/scene/text_input_hub.h

```diff
@@ -97,6 +98,48 @@ struct TextInputState
     std::optional<TextInputContentPurpose> content_purpose;
 };
 
+class CopyableWlArray
+{
+public:
+    explicit CopyableWlArray(wl_array* in_data)
+        : data_{copy(in_data)}
+    {
+    }
+
+    CopyableWlArray(CopyableWlArray const& other)
+        : data_{copy(other.data_)}
+    {
+    }
```

Comment:
This could be a forwarding constructor
```suggestion
    CopyableWlArray(CopyableWlArray const& other)
        : CopyableWlArray{other.data_}
    {
    }
```